### PR TITLE
dialects: (hw) Add `hw.module` and `hw.output` operations

### PR DESCRIPTION
**Pull request by Moxinilian**
This PR introduces the `hw.module` and its companion `hw.output` from CIRCT, while adding various small enhancements to the general infrastructure to accommodate them. ModuleOps are used to specify self-contained hardware modules, akin to functions in software.

A `hw.module` looks like the following:
```
hw.module @name<param1: i1, param2: i2>(in %a: i8, in %b: i9, out c: i8) {
  %res = arith.constant 0 : i8
  hw.output %res : i8
}
```
While hardware modules have similar semantics to functions, both inputs and outputs in `hw.module` are defined within the parenthesis, as *ports*. This is because some ports may be both input and output in CIRCT (currently not supported). Modules can optionally receive static parameters (in the example, `param1` and `param2`). Those are akin to C++ value templates or Rust const generics, and are useful to factor similar modules together.

`DirectionAttr`, `ModulePort` and `ModuleType` are defined to represent the type of a module and its ports. `ParamDeclAttr` is an attribute to represent the parameter definitions on the module.

The `hw.output` operation is the terminator for `hw.module`, setting the value of the output ports.

cc @superlopuh @math-fehr

### CODE REVIEW

File: xdsl/dialects/hw.py [hunk 1]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
             printer.print_string("]")
 
 
+class Direction(Enum):
+    """
+    Represents the direction of a module port.
+    """
+
+    # TODO: support INOUT direction
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:434]: Can you file an issue please and add it to the comment

File: xdsl/dialects/hw.py [hunk 2]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[Attribute]:
+        def parse_port() -> ModulePort:
+            direction = Direction.parse(parser)
+            if (
+                name := parser.parse_optional_identifier()
+                or parser.parse_optional_str_literal()
+            ) is None:
+                parser.raise_error(
+                    "expected port name as identifier or string litteral"
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:512]:
```suggestion
                    "expected port name as identifier or string literal"
```

File: xdsl/dialects/hw.py [hunk 3]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+@irdl_attr_definition
+class ParamDeclAttr(ParametrizedAttribute):
+    name = "hw.param.decl"
+
+    port_name: ParameterDef[StringAttr]
+    type: ParameterDef[TypeAttribute]
+
+    @classmethod
+    def parse_free_standing_parameters(
+        cls, parser: AttrParser, always_parse_name_in_quotes: bool = False
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:547]: Can you please document the behaviour of this argument?

Comment (superlopuh) [reply xdsl/dialects/hw.py:547]: Probably with an example of what it can and cannot parse

File: xdsl/dialects/hw.py [hunk 4]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+            name = parser.parse_optional_identifier()
+        if name is None:
+            parser.raise_error("expected parameter name")
+        parser.parse_punctuation(":")
+        typ = parser.parse_attribute()
+        if not isinstance(typ, TypeAttribute):
+            parser.raise_error("expected type attribute for parameter")
+
+        if parser.parse_optional_punctuation("=") is not None:
+            # TODO: support default values for parameters
```
Comment (luisacicolini) [inline xdsl/dialects/hw.py:564]: Should this be an issue too? CC: @superlopuh

File: xdsl/dialects/hw.py [hunk 5]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+    ) -> Sequence[Attribute]:
+        """
+        Parses the parameter declaration without the encompassing angle brackets.
+        """
+
+        name = parser.parse_optional_str_literal()
+        if name is None and not always_parse_name_in_quotes:
+            name = parser.parse_optional_identifier()
+        if name is None:
+            parser.raise_error("expected parameter name")
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:557]:
```suggestion
        if name is None:
            if always_parse_name_in_quotes:
                parser.raise_error("expected parameter name as string literal")
            name = parser.expect(parser.parse_optional_identifier, "expected parameter name")
```

File: xdsl/dialects/hw.py (hunk 4 above)
Comment (superlopuh) [inline xdsl/dialects/hw.py:564]: can you please file an issue

File: xdsl/dialects/hw.py [hunk 6]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+        parser.parse_punctuation(":")
+        typ = parser.parse_attribute()
+        if not isinstance(typ, TypeAttribute):
+            parser.raise_error("expected type attribute for parameter")
+
+        if parser.parse_optional_punctuation("=") is not None:
+            # TODO: support default values for parameters
+            parser.raise_error("default values for parameters are not yet supported")
+
+        return [StringAttr(name), typ]
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:567]: nit:
```suggestion
        return (StringAttr(name), typ)
```

File: xdsl/dialects/hw.py [hunk 7]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+                    f"module block argument (expected {port.type}, block argument "
+                    f"is of type {next_block_arg.type})"
+                )
+        if next(block_args, None) is not None:
+            raise VerifyException("too many block arguments in module block")
+
+    @classmethod
+    def parse(cls, parser: Parser) -> "ModuleOp":
+        @dataclass
+        class ModuleArg:
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:662]:
```suggestion
        class ModuleArg(NamedTuple):
```

Comment (superlopuh) [reply xdsl/dialects/hw.py:662]: These are a little simpler and have smaller overhead

**Review by luisacicolini — APPROVED** [review]
looks good to me!

File: xdsl/dialects/hw.py (hunk 1 above)
Comment (Moxinilian) [reply xdsl/dialects/hw.py:434]: #2368

File: xdsl/dialects/hw.py (hunk 4 above)
Comment (Moxinilian) [reply xdsl/dialects/hw.py:564]: #2367

File: xdsl/dialects/hw.py [hunk 8]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+
+    def verify_(self) -> None:
+        parent = self.parent_op()
+        assert isinstance(parent, ModuleOp)
+
+        expected_results = [
+            port.type
+            for port in parent.module_type.ports.data
+            if port.dir.data == Direction.OUTPUT
+        ]
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:812]: nit:
```suggestion
        expected_results = tuple(
            port.type
            for port in parent.module_type.ports.data
            if port.dir.data == Direction.OUTPUT
        )
```

File: xdsl/dialects/hw.py [hunk 9]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+                    f"output #{i} is of unexpected type (expected {expected}, got {got.type})"
+                )
+
+    @classmethod
+    def parse(cls, parser: Parser) -> "OutputOp":
+        operands = parser.parse_optional_undelimited_comma_separated_list(
+            parser.parse_optional_unresolved_operand, parser.parse_unresolved_operand
+        )
+        if operands is None:
+            return cls([])
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:831]: nit:
```suggestion
            return cls(())
```

File: xdsl/dialects/hw.py [hunk 10]
```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
...
+        printer.print(" ")
+        printer.print_list(self.inputs, printer.print_operand)
+        printer.print(" : ")
+        printer.print_list((x.type for x in self.inputs), printer.print_attribute)
+
+
 HW = Dialect(
     "hw",
-    [],
+    [ModuleOp, OutputOp],
```
Comment (superlopuh) [inline xdsl/dialects/hw.py:852]:
```suggestion
    [ModuleOp, OutputOp,],
```

File: xdsl/utils/lexer.py [hunk 11]
```diff
@@ -506,6 +506,7 @@ def lex(self) -> Token:
             f"Unexpected character: {current_char}",
         )
 
+    bare_identifier_regex = re.compile(r"[a-zA-Z_][a-zA-Z0-9_$.]*")
     bare_identifier_suffix_regex = re.compile(r"[a-zA-Z0-9_$.]*")
```
Comment (superlopuh) [inline xdsl/utils/lexer.py:510]: It seems like a good idea to do this in a separate PR, with explicit tests, probably reusing the suffix also
```suggestion
    SUFFIX = r"[a-zA-Z0-9_$.]*"
    bare_identifier_regex = re.compile(r"[a-zA-Z_]" + SUFFIX)
    bare_identifier_suffix_regex = re.compile(SUFFIX)
```

File: xdsl/printer.py [hunk 12]
```diff
@@ -226,12 +227,14 @@ def _print_results(self, op: Operation) -> None:
         self.print_list(op.results, self.print)
         self.print(" = ")
 
-    def print_ssa_value(self, value: SSAValue) -> None:
+    def print_ssa_value(self, value: SSAValue) -> str:
```
Comment (superlopuh) [inline xdsl/printer.py:230]: can you please do this in a separate PR also, with associated tests? Probably there are existing tests to update

**Review by superlopuh — APPROVED** [review]

File: xdsl/printer.py (hunk 12 above)
Comment (superlopuh) [reply xdsl/printer.py]: #2377

File: xdsl/dialects/hw.py [hunk 13]
```diff
@@ -409,12 +426,443 @@ def print_parameters(self, printer: Printer):
...
+        return direction
+
+    def print(self, printer: Printer, short: bool = False) -> None:
+        match self:
+            case Direction.INPUT:
+                printer.print("input" if not short else "in")
+            case Direction.OUTPUT:
+                printer.print("output" if not short else "out")
+
+    def is_input_like(self) -> bool:
```
Comment (lucjaulmes) [inline xdsl/dialects/hw.py:460]: Really feels like `is_input` at the moment. Is this meant to change when INOUT is supported? Ive seen a couple of `isInput() &` !isInOut()` in upstream code but not so much `isInput() ||isInOut()` so I’m curious where this is going.

**Review by lucjaulmes — COMMENTED** [review]
Thanks, nice contribution! A bit hard to get my head around all the various parts though. Here’s a few thoughts

File: xdsl/dialects/hw.py (hunk 13 above)
Comment (Moxinilian) [reply xdsl/dialects/hw.py:460]: Yeah this is the case. I used to go further into the inout implementation and this was useful, notably in the parser/printer to decide on whether it should be an SSA value or not.

File: xdsl/dialects/hw.py [hunk 14]
```diff
@@ -409,12 +426,443 @@ def print_parameters(self, printer: Printer):
             printer.print_string("]")
 
 
+class Direction(Enum):
+    """
+    Represents the direction of a module port.
+    """
+
+    # TODO: support INOUT direction (https://github.com/xdslproject/xdsl/issues/2368)
+    INPUT = (0,)
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py:435]: I think this is the result of formatting or something, when I wrote it it was indeed int values. I'll try to get rid of this.

File: xdsl/dialects/hw.py [hunk 15]
```diff
@@ -409,12 +426,443 @@ def print_parameters(self, printer: Printer):
...
+
+_MODULE_OP_ATTRS_HANDLED_BY_CUSTOM_FORMAT: list[str] = [
+    "sym_name",
+    "module_type",
+    "parameters",
+]
+
+
+@irdl_op_definition
+class ModuleOp(IRDLOperation):
```
[the remark this answers is not in the document]
Comment (lucjaulmes) [reply xdsl/dialects/hw.py]: The trait is just called `InnerRefNamespace` in the circt codebase – the suffix is added due to a name clash (there is also an `InnerRefNamsespace` class). The trait definition is in `include/circt/Dialect/HW/HWOpInterfaces.h` and `include/circt/Dialect/HW/HWOpInterfaces.td`. There’s some brief explanation about it in the “Traits and Classes” section of circt’s `docs/RationaleSymbols.md` document.

[the remark this answers is not in the document]
Comment (lucjaulmes) [reply xdsl/dialects/hw.py]: The operation that has that trait is the firrtl `CircuitOp` operation.

[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py]: Oh of course! Right. I'm just going to add it to the module operation, hopefully this makes sense.

[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py]: > The operation that has that trait is the firrtl CircuitOp operation.

That does not sound right, HwModules can live in builtin.module too. I think they may also be adding it to MLIR's ModuleOp via an external model, but I cannot find where this is.

[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py]: Ah no, I think the problem is in the InnerSymbolTable trait implementation in xDSL. It does not check for an InnerRefNamespaceLike but an InnerRefNamespace. I can probably fix that.

[the remark this answers is not in the document]
Comment (Moxinilian) [reply xdsl/dialects/hw.py]: Actually I don't understand what is happening in InnerSymbolTableTrait, so if you don't mind I just will not add the trait for now. My understanding is that the implementation of `InnerSymbolTableTrait` is not complete enough to be used anyway.

[the remark this answers is not in the document]
Comment (lucjaulmes) [reply xdsl/dialects/hw.py]: I think that’s fine, let’s not make this PR to large. I suppose the propoer way to add it later is to just add a comment and file a follow-up issue. LGTM for the rest.
